Disable write buffer in debug builds (M3/M4)

kjbracey opened the pull request:
### Description

As part of work to improve the debugging of exceptions, have
Mbed OS make an effort to make exceptions more precise in debug builds
at the cost of performance.

Related pyOCD work:

   https://github.com/mbedmicro/pyOCD/pull/430

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

kjbracey committed:
Disable write buffer in debug builds (M3/M4)

As part of work to improve the debugging of exceptions, have
Mbed OS make an effort to make exceptions more precise in debug builds
at the cost of performance.

Related pyOCD work:

   https://github.com/mbedmicro/pyOCD/pull/430

kjbracey commented:
I think my main desire for this (writes to NULL) is relieved by the latest version of #8335, but could still be worthwhile for other weird bus accesses.

NirSonnenschein commented:
CI is still problematic, but trying to test this for build/export 
/morph build

0xc0170 commented:
Info: This PR has been re-bundled into a new rollup PR (#8800).

No further work is needed here, as once that PR is merged, this PR will also be closed and marked as merged.
If any more commits are made in this PR, this PR will remain open and have to go through CI on its own.

0xc0170 commented:
This needs work. Exporters failure here are valid (latest results above) and were reproduced in the rollup

kjbracey commented:
Will investigate. You can treat this as low priority - it's a nice-to-have, but not really needed for 5.11 as long as #8335 makes it in.

kjbracey commented:
Okay, re-read the manual while investigating M3 revisions (turns out not all revisions of M3 have the bit, and not all platforms specify their M3 revision, and `core_cm3.h` is unhelpfully designed).

This bit only affects behaviour when the MPU is turned off, so it interacts weirdly with #8335, and there's no simple portable way to reproduce the behaviour with the MPU on, so I'm going to just drop it.

#8335's ROM write protection should cover 90% of the cases I'm worried about - will revisit with an MPU solution if I hit others. (Maybe make some areas Strongly Ordered in debug builds?)

kjbracey committed:
CM3: Make ACTLR bit definitions conditional

The ACTLR register itself is conditional on chip revision, but its
bit definitions were always defined.

Make the the bit definitions also conditional, so it is possible to
produce portable code that sets DISDEFWBUF if available:

    #ifdef SCnSCB_ACTLR_DISDEFWBUF_Msk
       SCnSCB->ACTLR |= SCnSCB_ACTLR_DISDEFWBUF_Msk;
    #endif

(cherry-picked from CMSIS b2b04dbeece0a046556bfc320bef6b20bef3f16f)

kjbracey commented:
Reopening, as #8335 does not in fact cover my most common case - the K64F.

Even if the K64F had a MPU driver, it still wouldn't provide precise aborts like the ARM MPU. So this patch is still useful.

Now incorporating a CMSIS patch to make the ACTLR register easier to deal with - should fix the previous CI failures. That has already been merged upstream: https://github.com/ARM-software/CMSIS_5/pull/478

kjbracey commented:
Shouldn't need work any more - I believe the extra commit clears the previous CI failures.

0xc0170 commented:
CI started
